### ISSUE TITLE
Typescript definition problems with apollo-client@next and link-core / cache-core

### ISSUE THREAD

**Issue by lordnox** [apollographql/apollo-client#2212]
![screen shot 2017-09-26 at 11 01 45](https://user-images.githubusercontent.com/503189/30852247-6b13b4e0-a2ab-11e7-86f0-83450aa05167.png)
Code in question:
```

import ApolloClient from 'apollo-client'
import HttpLink from 'apollo-link-http'
import InMemoryCache from 'apollo-cache-inmemory'

const link = new HttpLink({
  uri: 'https://graphql-pokemon.now.sh',
})

const cache = new InMemoryCache()

export const client = new ApolloClient({
  cache,
  link,
})

```

**Intended outcome:**
This code is mostly copied from `apollo-client-preset` and I expected it to work out of the box. I might be missing some parameters for the cache, but typescript did not like the code even without the cache

**Actual outcome:**
Typescript complains about:
- Types of property 'cache' are incompatible
Type 'InMemoryCache' is not assignable to type 'ApolloCache<{}>
Property 'write' is missing in type 'InMemoryCache'.

**How to reproduce the issue:**
The code in the beginning should produce the error

**Version**
- yarn@1.1.0
- node@8.5.0
- apollo-cache-inmemory@0.1.0-1
- apollo-client@2.0.0-0
- apollo-link-http@0.5.4
- typescript@2.5.2

**Comment by jbaxleyiii** [comment]
@lordnox can you try with the latest betas?

Also the `Cache` will need a type of the [store now](https://github.com/apollographql/apollo-client/blob/master/packages/apollo-cache/src/cache.ts#L8)

**Comment by katopz** [comment]
In my case
> package.json
```
  "dependencies": {
    "apollo-client-preset": "^0.2.0-beta.8",
    "react": "^16.0.0",
    "react-apollo": "^2.0.0-beta.0",
    "react-dom": "^16.0.0",
    "react-scripts-ts": "2.7.0",
    "redux": "^3.7.2"
  },
  "devDependencies": {
    "@types/jest": "^21.1.2",
    "@types/node": "^8.0.34",
    "@types/react": "^16.0.10",
    "@types/react-dom": "^16.0.1"
  }
```
> index.tsx
```
import * as React from 'react';
import { render } from 'react-dom';

import ApolloClient from 'apollo-client';
import { HttpLink, InMemoryCache } from 'apollo-client-preset';
import { ApolloProvider } from 'react-apollo';

import App from './App';

const GRAPHQL_END_POINT = 'SOME_ENDPOINT';

const client = new ApolloClient({
  link: new HttpLink({ uri: GRAPHQL_END_POINT }),
  cache: (new InMemoryCache()).restore({})
});

const WrappedApp = (
  <ApolloProvider client={client}>
    <App />
  </ApolloProvider>
);

render(WrappedApp, document.getElementById('root'));
```
> Error
```
(22,21): error TS2322: Type '{ client: ApolloClient<NormalizedCache>; children: Element; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<ApolloProvider> & Readonly<{ children?: ReactNode;...'.
  Type '{ client: ApolloClient<NormalizedCache>; children: Element; }' is not assignable to type 'Readonly<ProviderProps<Cache>>'.
    Types of property 'client' are incompatible.
      Type 'ApolloClient<NormalizedCache>' is not assignable to type 'ApolloClient<Cache>'.
        Type 'NormalizedCache' is not assignable to type 'Cache'.
          Property 'add' is missing in type 'NormalizedCache'.
```

**Comment by mzygmunt** [comment]
@katopz The same error with version apollo-client 2.0.1 and react-apollo 2.0.0

**Comment by lednhatkhanh** [comment]
This is really annoying, the old `subscriptions-transport-ws` version 0.8.3 has problems with Typescript so I decided to update to apollo client 2, now I got another problem with apollo-client... 
Have to stop and wait for the error to be fixed...

**Comment by Jannis** [comment]
This seems closely related (or identical even) to https://github.com/apollographql/react-apollo/issues/1299.

**Comment by jbaxleyiii** [comment]
Looks like someone took a stab at it [here](https://github.com/apollographql/react-apollo/pull/1319) which I'm working to release asap! Hopefully it fixes these issues!

**Comment by prencher** [comment]
Is there any traction on this? We just tried to upgrade our codebase today and ran into this with the latest versions.

**Comment by prencher** [comment]
Is there any traction on this? We just tried to upgrade our codebase today and ran into this with the latest versions.

@jbaxleyiii I don't see how the `ApolloProvider` change would address this. The issue is with `ApolloClient` and `InMemoryCache`.

**Comment by alvis** [comment]
@prencher the code in #1319 hasn't released yet, so that may be the reason the issue remains. May be you can try the master version for now. Disclaimer: I haven't tried it yet.

**Comment by iamclaytonray** [comment]
What is the current status on this? I just started refactoring my codebase to TypeScript and am running into the same issue.

**Comment by iamclaytonray** [comment]
@jbaxleyiii - Any word on when this will be released?

**Comment by katopz** [comment]
@iamclaytonray Seem working for me now, do try here https://github.com/katopz/apollo-client-preset-ts-example

**Comment by evenfrost** [comment]
Any update on this?

**Comment by iamclaytonray** [comment]
I can verify that the latest release fixes this problem. (Probably the release before latest as well. I haven't touched that project in about a couple of months.)

**Comment by hwillson** [comment]
This issue has been resolved (verified with `apollo-client` 2.3.1). Thanks!
